src/Editor.js

```diff
@@ -0,0 +1,127 @@
+import * as THREE from 'three';
+import { getCurrentTime } from './Util.js';
+
+export class Editor {
+  constructor(viewer, options = {}) {
+    this.viewer = viewer;
+
+    this.editPanel = null;
+
+    this.pointerUpHandlerPlaneFinder = this.onPointerUpPlaneFinder.bind(this);
+
+    this.initialized = false;
+    this.init();
+  }
+
+  init() {
+    if (this.initialized) return;
+
+    this.setupEditPanel();
+
+    this.initialized = true;
+  }
+
+  setupEditPanel() {
+    this.editPanel = document.createElement('div');
+    this.editPanel.style.position = 'absolute';
+    this.editPanel.style.padding = '10px';
+    this.editPanel.style.backgroundColor = '#cccccc';
+    this.editPanel.style.border = '#aaaaaa 1px solid';
+    this.editPanel.style.zIndex = 90;
+    this.editPanel.style.width = '375px';
+    this.editPanel.style.fontFamily = 'arial';
+    this.editPanel.style.fontSize = '10pt';
+    this.editPanel.style.textAlign = 'left';
+
+
+    const editTable = document.createElement('div');
+    editTable.style.width = '100%';
+
+
+    // Plane Finder
+    const planeFinder = document.createElement('div');
+    planeFinder.style.width = '100%';
+    planeFinder.style.display = 'flex';
+    planeFinder.style.flexDirection = 'row';
+    planeFinder.style.justifyContent = 'space-between';
+
+    const planeFinderLabel = document.createElement('p');
+    planeFinderLabel.id = 'planeFinderLabel';
+    planeFinderLabel.innerHTML =
+      `Up: ${this.viewer.camera.up.x.toFixed(3)}, ${this.viewer.camera.up.y.toFixed(3)}, ${this.viewer.camera.up.z.toFixed(3)}`;
+
+    const planeFinderButton = document.createElement('button');
+    planeFinderButton.innerHTML = 'Find ground plane';
+    planeFinderButton.addEventListener('click', () => {
+      this.setupPlaneFinder();
+    });
+
+    planeFinder.appendChild(planeFinderLabel);
+    planeFinder.appendChild(planeFinderButton);
+
+    editTable.appendChild(planeFinder);
+
+    this.editPanel.appendChild(editTable);
+    this.editPanel.style.display = 'block';
+    this.viewer.renderer.domElement.parentElement.prepend(this.editPanel);
+  }
+
+  setupPlaneFinder() {
+    if (this.viewer.useBuiltInControls) {
+      this.viewer.rootElement.removeEventListener('pointerup', this.viewer.pointerUpHandler);
```

Comment:
I definitely think we need a less intrusive way of overriding behavior in `Viewer`. I'm still not sure we should simply have `Editor` inherit from `Viewer` and just override the `onMouseDown()` etc. functions, or if we should add functions to `Viewer` to allow them to be set by an external source.

Reply:
I'm mainly a React developer and am not all that familiar with best practices in vanilla and OOP.This could be a case of lack of knowledge on my part.

The effect I was going for was to have a default global state that is the same as just the viewer. I couldn't really figure out how to do it with inheritance + overriding without loosing the default state. In my mind we had to store those handlers away anyhow and I figured that we might aswell keep the class itself intact as that stored default state.

Reply:
So I was thinking something along the lines of:
```javascript
class Viewer {

    constructor() {
        window.addEventListener("mousedown", this.onMouseDown.bind(this));
    }

    onMouseDown() {
        // default behavior
    }
}


const PointerMode = {
    Default: 0,
    CameraUpSelection: 1
};

class Editor extends Viewer {

    constructor() {
        super();
        this.pointerMode = PointerMode.Default;
        this.setupEditPanel();
    }

    setupEditPanel() {
        const planeFinderButton = document.createElement('button');
        planeFinderButton.innerHTML = 'Find ground plane';
        planeFinderButton.addEventListener('click', () => {
            this.pointerMode = PointerMode.CameraUpSelection;
        });
    }

    // override Viewer.onMouseDown()
    onMouseDown() {
        if (this.pointerMode === PointerMode.CameraUpSelection) {
            // Do custom stuff for camera up selection
        } else {
            super.onMouseDown();
        }
    }  

}
```
This would retain default behavior when not selecting the camera-up vector. (This code probably contains errors, but hopefully it conveys the idea :) )

Reply:
Yes, this is absolutely more clear, I like it.

With this structure I would prefer to inherit the Viewer class. 

I also think a user should be able to jack into the events and extend the functionality in a straight forward manner. I'm not really sure how one would do that right now, maybe something like how I bound the invalidate() method to the controllers change event. Is there something with this approach that would hinder that?

Reply:
I don't think an inheritance based approach will hinder that; we can still extend functions in `Viewer` by overriding them with custom functionality and making sure to call the base function along with it. We have more options that way because we can override any function and it's cleaner because it doesn't require changes to `Viewer`. I do agree that it would be nice for `Viewer` to have a general way to hook into events so that anyone developing with the library could hook into them. Maybe we need to make `Viewer` extend an `EventDispatcher` class and emit events at the appropriate places.

Reply:
+1 on making `Viewer` extend `EventDispatcher`